Iss10 - Simple log in user story

aThorp96 opened the pull request:
The branch satisfies all points in the issue. I think the servercom package could be used for the majority of the backend functions.

aThorp96 committed:
Structure in place, but import not wokring

aThorp96 committed:
Import statement still not working

aThorp96 committed:
Structure in place, but import statement not working

aThorp96 committed:
Import statement working. Login not yet working

aThorp96 committed:
Fixed outdated API spike

aThorp96 committed:
Successfully logs user in and notifies the user connectivity status.

aThorp96 committed:
Delete .loginTest.go.swp

aThorp96 committed:
Merge branch 'master' into iss10

aThorp96 committed:
Fixed travis issue

aThorp96 committed:
Merge branch 'iss10' of https://github.com/LinuxAtApp/matterleast into iss10

aThorp96 committed:
Modified main.go to satisfy main_test.go

whereswaldon commented on main.go:
```diff
@@ -1,9 +1,26 @@
+	flag.Parse()
+	url := flag.Arg(0)
+	//Creates client and logs user in.
+	client:= *servercom.Startup(url, *username, *password)
```
@aThorp96 Why are you de-referencing the pointer that `servercom.Startup()` returns?

aThorp96 replied:
I didn't do that intentionally.

whereswaldon commented on main.go:
```diff
@@ -1,9 +1,26 @@
+	//Creates client and logs user in.
+	client:= *servercom.Startup(url, *username, *password)
+	//Tests if login was successful. 
+	if servercom.Connected(client) {
```
Maybe it would be worthwhile to wrap the `*mm.Client` that we get back in our own structure so that we can attach custom methods to it. That way, we could test this more cleanly with something like `if client.Connected() {...}`?

aThorp96 replied:
That is something I wanted to do. I wasn't sure the right way to go about it with my limited knowledge of go. I'll look into the using a struct.

whereswaldon replied:
Specifically, try struct embedding. It would look something like this:
```go
type Server struct {
    *mm.Client
}
```
Let's say that we had a `*servercom.Server` called `s`. We could access the client with `s.Client`. However, we could also call custom methods like `s.Connected()`.

whereswaldon commented on servercom/servercom.go:
```diff
@@ -0,0 +1,27 @@
+package servercom
```
This package does everything that I mentioned, well done! Now I think we should look at how to make it clean/easy to use from the outside. As I mentioned above, I think it might be better to define our own `struct` that has a `*mm.client` as a field. This would allow us to write custom methods that add a level of convenience on top of the basic client that mattermost wrote for us. Thoughts?

whereswaldon reviewed (commented):
Solid job! This has all of the functionality that I was looking for. I did have some design questions for you though. (See code for details)

aThorp96 committed:
Added ServerCom struct and changed function  into a ServerCom method.

aThorp96 committed:
Added some more features to ServerCom. Now supports setting the teamId and can make a post when given a string message

aThorp96 committed:
Added channel and team fields in ServerCom package. Added accessor, modifier and print methods to new fields. Newpost method now working.

aThorp96 committed:
Added many more methods to ServerCom package.

aThorp96 commented:
The ServerCom package now offers all the functionality of the API spike. I think it might be better to store the team and maybe channel field in the config file so you don't have to select one each time you run the program. Other than that I think there's not a lot to add to the package for now.

aThorp96 committed:
Attempt at fixing issue with travis tests

aThorp96 committed:
fixed incorrect documentation

aThorp96 commented:
@camwood23 These methods work?

whereswaldon commented on servercom/servercom.go:
```diff
@@ -0,0 +1,138 @@
+/*
+Startup accepts the url and login credentials for a user, and returns a new serverCom struct.
+*/
+func Startup(url string, username string, password string) ServerCom {
```
I have a couple of recommendations here. Since this method is used to construct `ServerCom` structs, and since all of your methods are defined on `*ServerCom` pointers, you might as well return a pointer here.

On another note, error handling in library functions can be counterintuitive. I quite like [this guy's thoughts](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) on the subject. At the least, I would recommend changing the return types of most of these functions to return an error in addition to their normal return type. Then, instead of printing the error, return it. 
The reason for this is that you don't really want library functions dumping things onto Stdout. That makes it very hard to control what is printed to the screen from the `main()` function of your application. If you return errors to main, then main can decide to print them (or not to) at its discretion. Does that make sense?

aThorp96 replied:
That makes sense. I had some of the earlier methods return errors initially. I changed them to print instead when I noticed the setters didn't actually return any value, and i thought it might be weird to have something like `err :=  MethodCall()`. But it makes sense having the freedom to chose how to handle them.

whereswaldon replied:
Yeah, I understand about those setters, but it's the convention. Go prefers explicit error handling over the `panic/recover` mechanism for errors that can be predicted.

camwood23 replied:
Sorry for the delayed response, but yeah I do like these methods. However, we also need some way to tell whenever a new message is available. The screen only updates whenever an event happens, so if there is a new message that was sent, you won't see the message until you press a key.

There are two ways that I think we can handle this though, we can either have a timer where the screen gets updated every so often or we can have the server send everyone's app a message saying that there is a new message available and the screen needs to be updated to reflect that change.

Also is there a way to get a list of all the other people connected to the chat? I was thinking we could use that information to assign random colors as people connect and disconnect. I'm not sure if that's what one of the team methods does or not?

aThorp96 replied:
A package running in parallel to the interface that just compared the length of the message list to a variable storing the previous length of the message length makes the most sense to me. If the length is greater then it updates the variable and returns non-nil.

aThorp96 replied:
@camwood23 I'm sure there is a way to do that. I'll look into it.

whereswaldon replied:
@aThorp96 @camwood23 [This code](https://github.com/mattermost/mattermost-bot-sample-golang/blob/master/bot_sample.go#L82) shows how to detect events on the server. Not sure how much adaptation it might need to work for us, but it's a better method than polling.

whereswaldon commented on servercom/servercom.go:
```diff
@@ -0,0 +1,138 @@
+
+//GetChannels returns a map with all availible channels in team.
+//**Must have successfully run SetTeam()**
+func (sc *ServerCom) GetChannels() *mm.ChannelList {
```
Maybe this should explicitly check whether `SetTeam` has been called. If not, you could return a custom error using `fmt.Errorf()`

aThorp96 replied:
I like that. I'll look into custom errors

whereswaldon commented on servercom/servercom.go:
```diff
@@ -0,0 +1,138 @@
+	//Extracts Posts in order into postsSLice
+	postSinceDate := postSinceDateResult.Data.(*mm.PostList)
+	postsSlice := make([]*mm.Post, len(postSinceDate.Order))
+	for index, key := range postSinceDate.Order {
```
Why do we need to convert this map into a slice? I think I'm missing something here.

aThorp96 replied:
With a slice we don't have to create a counter variable, since we use index within the loop. Though I thought there was a way to just append a slice I didn't have any luck finding it. I suppose it might be better to have an int variable instead of a whole new list of post objects.

whereswaldon commented on servercom/servercom.go:
```diff
@@ -0,0 +1,138 @@
+//Postcount is the number of posts before (and including) the offset that will be in the slice.
+func (sc *ServerCom) GetSelectPosts(offset int, postCount int) []*mm.Post{
+	postList := sc.GetChannelData()
+	selectPosts := make([]*mm.Post, postCount)
```
There's a cool trick with slices that can save you a ton of time here. Your code from here down is (I think) equivalent to saying:
```go
return postList[offset:offset+postCount]
```
See [here](https://gobyexample.com/slices) for details.

whereswaldon reviewed (changes requested):
Wow, @aThorp96, this code is really coming along! I think we should update the methods to return errors, but that's the only significant wrinkle in here. There's one other thing that I think might be worth considering though: How can we test this code? (Can we test this code?) Automated tests are a good thing, so can you see a way to write them for your ServerCom code?

whereswaldon commented:
@aThorp96 What's the word here?

aThorp96 commented:
@whereswaldon I've been too busy with school and work to contribute. I might be able to do a little this weekend but I can't say for sure.

whereswaldon commented:
Would you like someone else to take over this branch for a while? I haven touched it because I thought that I might step on your toes.

> On Jun 5, 2017, at 1:35 PM, Andrew Thorp <notifications@github.com> wrote:
> 
> @whereswaldon I've been too busy with school and work to contribute. I might be able to do a little this weekend but I can't say for sure.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

aThorp96 commented:
@Whereswaldon that might be best. I'll push what I've done this afternoon.

aThorp96 committed:
Not yet working, needs a few kinks ironed out. I din't have time to work on it durring the summer so I'm just pushing what changes I've made.

whereswaldon committed:
Ensure all functions consistently return errors

whereswaldon committed:
Ensure that main compiles

whereswaldon committed:
Update error hanlder to more semantic name

whereswaldon committed:
Add team selection flag

whereswaldon committed:
Add channel selection flag

whereswaldon committed:
Replace specific error type with 'error'

whereswaldon committed:
Make servercom output configurable

whereswaldon committed:
Have fatal actually kill program

whereswaldon committed:
Reformat code

whereswaldon committed:
Simplify slice operation

whereswaldon committed:
Implement reverse-chronological channel dump

whereswaldon committed:
Fix index-out-of-bounds problems

whereswaldon committed:
Create interface to wrap *mm.Client for testing

whereswaldon committed:
Remove Connected method from servercom

The connected method was redundant. We would have returned an error
during a prior call to Startup() if there were a problem logging in.
It also relied on a Field of *mm.Client, rather than a method, which
made it hard to wrap *mm.Client in an interface.

whereswaldon committed:
Fully wrap *mm.Client in an interface for testability

whereswaldon committed:
Create mock client implementation for testing

NaturalFractals commented on servercom/servercom_test.go:
```diff
@@ -0,0 +1,49 @@
+
+func (m *mockClient) GetPostsSince(channelId string, time int64) (*mm.Result, *mm.AppError) {
+	return m.NextResult, m.NextErr
+}
```
I like how all of these function are handled, looks very clean.

NaturalFractals commented on servercom/servercom.go:
```diff
@@ -0,0 +1,157 @@
+/*
+Startup accepts the url and login credentials for a user, and returns a new serverCom struct.
+*/
+func Startup(url string, username string, password string, output io.Writer) (*ServerCom, error) {
```
Is this an internal method only? 

I know were are just creating spikes, and this is more of a general question, but is it considered best practice to use **camelCase** for internal methods and **CamelCase** for public methods?

Does go fmt already take care of this?

whereswaldon replied:
No, this is  essentially the constructor for a `*ServerCom`. The convention is to use camelCase for private functions and PascalCase for public ones. It's not really a convention though, because it's enforced by the compiler. This isn't something that `go fmt` takes care of because it radically affects the code's logic.

NaturalFractals commented on servercom/servercom.go:
```diff
@@ -0,0 +1,157 @@
+    GetChannels(etag string) (*mm.Result, *mm.AppError)
+    CreatePost(post *mm.Post) (*mm.Result, *mm.AppError)
+    GetPostsSince(channelId string, time int64) (*mm.Result, *mm.AppError)
+}
```
Looks like a great interface! Seems as though this would make easy to add any additional functionality in the future.

whereswaldon commented:
Best practice is to camelCase private fields and variables and to PascalCase public fields and variables

This function is essentially the constructor, but it would be a good idea to name it better

> On Jun 19, 2017, at 9:38 PM, Jesse Cochran <notifications@github.com> wrote:
> 
> @NaturalFractals commented on this pull request.
> 
> In servercom/servercom.go:
> 
> > +var _ Client = &mm.Client{}
> +
> +/*
> +ServerCom acts as a mitigator between the frontend and the mattermost model API.
> +*/
> +type ServerCom struct {
> +	Output  io.Writer // This is where the print methods will send output
> +	Client  Client
> +	Team    mm.Team
> +	Channel *mm.Channel
> +}
> +
> +/*
> +Startup accepts the url and login credentials for a user, and returns a new serverCom struct.
> +*/
> +func Startup(url string, username string, password string, output io.Writer) (*ServerCom, error) {
> Is this an internal method only?
> 
> I know were are just creating spikes, and this is more of a general question, but is it considered best practice to usecamelCase for internal methods and CamelCase for public methods?
> 
> Does go fmt already take care of this?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

whereswaldon committed:
Remove print functionality

whereswaldon committed:
Implement a crude version of printing the channels and teams in main

whereswaldon committed:
Wrap the websocket client in an interface

whereswaldon committed:
Create factory functions for client interfaces

whereswaldon committed:
Wrap client struct to expose fields as interface methods

whereswaldon committed:
Implemented websocket-based event listening

whereswaldon committed:
Clean up the output
